astropy/io/fits/hdu/compressed/compressed.py

```diff
@@ -572,408 +526,146 @@ def data(self):
         not need to access the whole array, consider instead using the
         :attr:`~astropy.io.fits.CompImageHDU.section` property.
         """
-        if len(self.compressed_data) == 0:
+        # If there is no internal binary table, the HDU was not created from a
+        # file and therefore the data is just the one on the parent ImageHDU
+        # class
+
+        if not self._decompression_active:
+            return super().data
+        elif len(self._bintable.data) == 0:
             return None
 
-        # Since .section has general code to load any arbitrary part of the
-        # data, we can just use this - and the @lazyproperty on the current
-        # property will ensure that we do this only once.
-        data = self.section[...]
+        if getattr(self, "_decompressed_data", None) is not None:
+            return self._decompressed_data
 
-        # Right out of _ImageBaseHDU.data
-        self._update_header_scale_info(data.dtype)
+        # Since .section has general code to load any arbitrary part of the
+        # data, we can just use this
+        data = self._decompressed_data = self.section[...]
```

Comment:
The goal here was to distinguish between the case where the data is actually set, which will require re-compression, and the case where the data is decompressed but unmodified. I thought it would be clearer this way which 'mode' we were in. In fact we wouldn't be able to use ``_has_data`` from ``BaseHDU`` because it accesses ``self.data`` which would force the decompression. I think we'd need to use ``_data_loaded``.

Would you mind if we kept it as-is for now since this is just internal details, and I can try and investigate in a follow-up PR how easy it would be to write it with ``_data_loaded`` and it will be easier to see at a glance if it improves things or makes things more confusing?

Reply:
> The goal here was to distinguish between the case where the data is actually set, which will require re-compression, and the case where the data is decompressed but unmodified

But once the full array is decompressed there is no way to know is it is modified by an external reference. 

I see now that you also removed the `lazyproperty`, in favor of the manual cache with `self._decompressed_data`. We can keep for later but it's just that I don't see why we need this. For me once the full array is decompressed, we no more need the bintable and the object should behave the same as if created from a numpy array ?

Reply:
Good point, I'm happy to update things to force decompression if the data is decompressed at any point.